Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp

```diff
@@ -2857,4 +2914,63 @@ DestinationColorSpace CanvasRenderingContext2DBase::colorSpace() const
     return toDestinationColorSpace(m_settings.colorSpace);
 }
 
+ExceptionOr<void> CanvasRenderingContext2DBase::beginLayer(const BeginLayerOptions&)
+{
+    // FIXME: apply filter.
+
+    FloatRect boundingBox { { }, canvasBase().size() };
+
+    auto* ctx = drawingContext();
+    if (!ctx) {
+        // FIXME: throw exceptions or not.
+        return { };
+    }
+
+    save();
+    realizeSaves();
+
+    ctx->beginTransparencyLayer(state().globalAlpha);
+
+    // Reset layer rendering states to default value: globalAlpha, globalCompositeOperation, shadowOffsetX, shadowOffsetY, shadowBlur, shadowColor.
+    setGlobalAlpha(1.0);
+    setGlobalCompositeOperation("source-over"_s);
+    setShadowOffsetX(0);
+    setShadowOffsetY(0);
+    setShadowBlur(0);
+    setShadowColor("black"_s);
+    modifiableState().isLayer = true;
+
+    m_layerCount++;
+
+    return { };
+}
+
+ExceptionOr<void> CanvasRenderingContext2DBase::endLayer()
+{
+    if (m_stateStack.isEmpty())
+        return Exception { InvalidStateError };
+
+    auto* ctx = drawingContext();
+    if (!ctx) {
+        // FIXME: what if drawingContext() is null.
+        return { };
+    }
+
+    realizeSaves();
+
+    if (!state().isLayer)
+        return Exception { InvalidStateError };
+
+    ctx->endTransparencyLayer();
+
+    modifiableState().isLayer = false;
+    restore();
+
+    m_layerCount--;
+
+    didDrawEntireCanvas();
```

Comment:
maybe should just call didDraw() with the clipping region prior to beginLayer()? need tests